File: android/app/src/main/java/app/ezber/android/models/Preset.kt
package app.ezber.android.models

/**
 * A saved drill definition: surah and section, per-verse repeats, reciter and
 * display options. Field-for-field the `presets` row in
 * schema/user_schema.sql, plus the optional per-verse overrides from
 * `preset_verses`.
 *
 * `id == 0L` means "not saved yet"; the store assigns the row id on insert.
 */
data class Preset(
    val id: Long = 0L,
    val name: String,
    val surahId: Int?,
    val range: VerseRange,
    val repeats: RepeatPlan = RepeatPlan(),
    val reciterId: Int? = null,
    val transliterationId: Int? = null,
    val translationId: Int? = null,
    val display: DisplayOptions = DisplayOptions(),
    val downloadScope: DownloadScope = DownloadScope.INHERIT,
    val createdAt: String? = null,
    val updatedAt: String? = null,
    val lastUsedAt: String? = null,
) {
    val repeatSummary: String
        get() = if (repeats.overrides.isEmpty()) {
            "${repeats.defaultRepeats}× each"
        } else {
            "${repeats.defaultRepeats}× each · ${repeats.overrides.size} override" +
                if (repeats.overrides.size == 1) "" else "s"
        }

    val totalRepetitions: Int get() = repeats.totalRepetitions(range)
}

/**
 * How many times each verse repeats. [overrides] holds per-verse exceptions to
 * [defaultRepeats]; the store writes them to `preset_verses`.
 */
data class RepeatPlan(
    val defaultRepeats: Int = 5,
    val overrides: Map<Int, Int> = emptyMap(),
) {
    fun repeatsFor(verseNumber: Int): Int = maxOf(1, overrides[verseNumber] ?: defaultRepeats)

    fun totalRepetitions(range: VerseRange): Int {
        if (range.isEmpty) return 0
        return (range.start..range.end).sumOf { repeatsFor(it) }
    }

    /** Immutable edit used by the preset builder: 1..99, default removes the override. */
    fun withRepeats(value: Int, verseNumber: Int): RepeatPlan {
        val clamped = value.coerceIn(1, 99)
        val updated = overrides.toMutableMap()
        if (clamped == defaultRepeats) {
            updated.remove(verseNumber)
        } else {
            updated[verseNumber] = clamped
        }
        return copy(overrides = updated)
    }
}

File: ios/Ezber/Models/Preset.swift
import Foundation

/// A saved drill definition: surah and section, per-verse repeats, reciter and
/// display options.
struct Preset: Identifiable, Hashable, Codable {
    var id: UUID
    var name: String
    var surahID: Int
    var range: VerseRange
    var repeats: RepeatPlan
    var reciterID: String
    var display: DisplayOptions
    var createdAt: Date
    var updatedAt: Date
    var lastStudiedAt: Date?

    init(
        id: UUID = UUID(),
        name: String,
        surahID: Int,
        range: VerseRange,
        repeats: RepeatPlan = RepeatPlan(),
        reciterID: String,
        display: DisplayOptions = DisplayOptions(),
        createdAt: Date = Date(),
        updatedAt: Date = Date(),
        lastStudiedAt: Date? = nil
    ) {
        self.id = id
        self.name = name
        self.surahID = surahID
        self.range = range
        self.repeats = repeats
        self.reciterID = reciterID
        self.display = display
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.lastStudiedAt = lastStudiedAt
    }

    var repeatSummary: String {
        if repeats.overrides.isEmpty {
            return "\(repeats.defaultRepeats)× each"
        }
        return "\(repeats.defaultRepeats)× each · \(repeats.overrides.count) override\(repeats.overrides.count == 1 ? "" : "s")"
    }

    var totalRepetitions: Int {
        repeats.totalRepetitions(in: range)
    }
}

/// How many times each verse repeats. `overrides` holds per-verse exceptions to
/// `defaultRepeats`.
struct RepeatPlan: Hashable, Codable {
    var defaultRepeats: Int
    var overrides: [Int: Int]

    init(defaultRepeats: Int = 5, overrides: [Int: Int] = [:]) {
        self.defaultRepeats = max(1, defaultRepeats)
        self.overrides = overrides
    }

    func repeats(forVerse verseNumber: Int) -> Int {
        max(1, overrides[verseNumber] ?? defaultRepeats)
    }

    func totalRepetitions(in range: VerseRange) -> Int {
        guard !range.isEmpty else { return 0 }
        return (range.start...range.end).reduce(0) { $0 + repeats(forVerse: $1) }
    }

    mutating func setRepeats(_ value: Int, forVerse verseNumber: Int) {
        let clamped = max(1, min(99, value))
        if clamped == defaultRepeats {
            overrides.removeValue(forKey: verseNumber)
        } else {
            overrides[verseNumber] = clamped
        }
    }

    mutating func resetOverrides() {
        overrides.removeAll()
    }
}
